From 25de2ba764fdcaff71f04dc5ed11f916231f1f3b Mon Sep 17 00:00:00 2001
From: Liu Ying <victor.liu@nxp.com>
Date: Fri, 7 Jul 2017 17:32:43 +0800
Subject: [PATCH 2085/5242] MLK-15925-1 gpu: imx: dpu: fetchdecode: Add
 get/set display stream id support

commit  6cb3c8e28b611c25892ba32b7d73147d66bccd6f from
https://source.codeaurora.org/external/imx/linux-imx.git

This patch adds two helpers so that users may get or set the display
stream id of fetchdecode.

Signed-off-by: Liu Ying <victor.liu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/imx/dpu/dpu-fetchdecode.c |   22 ++++++++++++++++++++++
 include/video/dpu.h                   |   10 ++++++++++
 2 files changed, 32 insertions(+)

diff --git a/drivers/gpu/imx/dpu/dpu-fetchdecode.c b/drivers/gpu/imx/dpu/dpu-fetchdecode.c
index 39f9cba..69f03fa 100644
--- a/drivers/gpu/imx/dpu/dpu-fetchdecode.c
+++ b/drivers/gpu/imx/dpu/dpu-fetchdecode.c
@@ -74,6 +74,8 @@ struct dpu_fetchdecode {
 	struct dpu_soc *dpu;
 	fetchtype_t fetchtype;
 	shadow_load_req_t shdlreq;
+	/* see DPU_PLANE_SRC_xxx */
+	unsigned int stream_id;
 };
 
 static inline u32 dpu_pec_fd_read(struct dpu_fetchdecode *fd,
@@ -414,6 +416,26 @@ shadow_load_req_t fetchdecode_to_shdldreq_t(struct dpu_fetchdecode *fd)
 }
 EXPORT_SYMBOL_GPL(fetchdecode_to_shdldreq_t);
 
+unsigned int fetchdecode_get_stream_id(struct dpu_fetchdecode *fd)
+{
+	return fd->stream_id;
+}
+EXPORT_SYMBOL_GPL(fetchdecode_get_stream_id);
+
+void fetchdecode_set_stream_id(struct dpu_fetchdecode *fd, unsigned int id)
+{
+	switch (id) {
+	case DPU_PLANE_SRC_TO_DISP_STREAM0:
+	case DPU_PLANE_SRC_TO_DISP_STREAM1:
+	case DPU_PLANE_SRC_DISABLED:
+		fd->stream_id = id;
+		break;
+	default:
+		WARN_ON(1);
+	}
+}
+EXPORT_SYMBOL_GPL(fetchdecode_set_stream_id);
+
 struct dpu_fetchdecode *dpu_fd_get(struct dpu_soc *dpu, int id)
 {
 	struct dpu_fetchdecode *fd;
diff --git a/include/video/dpu.h b/include/video/dpu.h
index 4463afe..32c77ea 100644
--- a/include/video/dpu.h
+++ b/include/video/dpu.h
@@ -428,6 +428,8 @@ void fetchdecode_yuv_constantcolor(struct dpu_fetchdecode *fd,
 void fetchdecode_controltrigger(struct dpu_fetchdecode *fd, bool trigger);
 int fetchdecode_fetchtype(struct dpu_fetchdecode *fd, fetchtype_t *type);
 shadow_load_req_t fetchdecode_to_shdldreq_t(struct dpu_fetchdecode *fd);
+unsigned int fetchdecode_get_stream_id(struct dpu_fetchdecode *fd);
+void fetchdecode_set_stream_id(struct dpu_fetchdecode *fd, unsigned int id);
 struct dpu_fetchdecode *dpu_fd_get(struct dpu_soc *dpu, int id);
 void dpu_fd_put(struct dpu_fetchdecode *fd);
 
@@ -490,6 +492,14 @@ void layerblend_pixengcfg_clken(struct dpu_layerblend *lb,
 struct dpu_tcon *dpu_tcon_get(struct dpu_soc *dpu, int id);
 void dpu_tcon_put(struct dpu_tcon *tcon);
 
+/*
+ * to avoid on-the-fly/hot plane resource migration
+ * between two display interfaces
+ */
+#define DPU_PLANE_SRC_TO_DISP_STREAM0	BIT(0)
+#define DPU_PLANE_SRC_TO_DISP_STREAM1	BIT(1)
+#define DPU_PLANE_SRC_DISABLED		0
+
 #define MAX_FD_NUM	4
 #define MAX_LB_NUM	7
 struct dpu_plane_res {
-- 
1.7.9.5

